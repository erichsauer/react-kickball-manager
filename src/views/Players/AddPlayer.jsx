import { useEffect, useState } from 'react'
import { Link, useHistory } from 'react-router-dom'
import PlayerForm from '../../components/players/PlayerForm'
import { getPlayerById, createPlayer } from '../../services/players'

function AddPlayer({
  user,
  match: {
    params: { id },
  },
}) {
  const [input, setInput] = useState({ name: '', position: '', teamId: '' })
  const [team, setTeam] = useState({ name: '', position: '', teamId: '' })
  const [alert, setAlert] = useState('')
  const history = useHistory()

  useEffect(() => {
    if (!user) history.push('/players')
  }, [user, history])

  useEffect(() => {
    setInput(team)
  }, [team])

  useEffect(() => {
    getPlayerById(id)
      .then(({ data }) => setTeam(data))
      .catch((err) => console.error(err))
  }, [id])

  const onChange = ({ target }) => {
    setInput((prevState) => ({ ...prevState, [target.name]: target.value }))
  }

  const onSubmit = async (event) => {
    event.preventDefault()

    const { data, error } = await createPlayer(input)

    if (error) return setAlert(error.message)
    return history.push(`/players/${data[0].id}`)
  }

  return (
    <>
      <Link to="/players">Players</Link> &raquo; <Link to={`/players/${id}`}>{team.name}</Link>{' '}
      &raquo; <span>Add</span>
      <br />
      <br />
      <p>{alert}</p>
      <PlayerForm
        label="Add Player"
        name={input.name}
        position={input.position}
        teamId={input.teamId}
        onChange={onChange}
        onSubmit={onSubmit}
      />
    </>
  )
}

export default AddPlayer
