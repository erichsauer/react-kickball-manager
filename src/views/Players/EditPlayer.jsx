import { useEffect, useState } from 'react'
import { Link, useHistory } from 'react-router-dom'
import PlayerForm from '../../components/players/PlayerForm'
import { getPlayerById, updatePlayerById } from '../../services/players'

function EditPlayer({
  user,
  match: {
    params: { id },
  },
}) {
  const [input, setInput] = useState({ name: '', position: '' })
  const [player, setPlayer] = useState({ name: '', position: '' })
  const [alert, setAlert] = useState('')
  const history = useHistory()

  useEffect(() => {
    if (!user) history.push('/players')
  }, [user, history])

  useEffect(() => {
    setInput(player)
  }, [player])

  useEffect(() => {
    getPlayerById(id)
      .then(({ data }) => setPlayer(data))
      .catch((err) => console.error(err))
  }, [id])

  const onChange = ({ target }) => {
    setInput((prevState) => ({ ...prevState, [target.name]: target.value }))
  }

  const onSubmit = async (event) => {
    event.preventDefault()

    const { data, error } = await updatePlayerById(player.id, {
      name: input.name,
      position: input.position,
    })

    if (error) return setAlert(error.message)
    return history.push(`/players/${data[0].id}`)
  }

  return (
    <>
      <Link to="/players">Players</Link> &raquo; <Link to={`/players/${id}`}>{player.name}</Link>{' '}
      &raquo; <span>Edit</span>
      <br />
      <br />
      <p>{alert}</p>
      <PlayerForm
        label="Edit Player"
        name={input.name}
        position={input.position}
        onChange={onChange}
        onSubmit={onSubmit}
      />
    </>
  )
}

export default EditPlayer
