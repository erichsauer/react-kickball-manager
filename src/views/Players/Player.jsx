import { useEffect, useState } from 'react'
import { Link } from 'react-router-dom'
import Loading from '../../components/loading/Loading'
import { getPlayerById } from '../../services/players'

function Player({
  user,
  match: {
    params: { id },
  },
}) {
  const [player, setPlayer] = useState({})
  const [loading, setLoading] = useState(true)

  useEffect(() => {
    getPlayerById(id)
      .then(({ data }) => setPlayer(data))
      .catch((err) => console.error(err))
      .finally(() => setLoading(false))
  }, [id])

  if (loading) return <Loading />

  return (
    <>
      <Link to="/players">Back to Players</Link>
      <h1>{player.name}</h1>
      <p>
        {player.position} for the {player.teams.name}
      </p>
      {user && (
        <p>
          <Link to={`/players/${id}/edit`}>Edit Player</Link>
        </p>
      )}
    </>
  )
}

export default Player
