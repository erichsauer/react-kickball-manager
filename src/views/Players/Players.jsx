import { useEffect, useState } from 'react'
import { Link } from 'react-router-dom'
import Loading from '../../components/loading/Loading'
import { deletePlayerById, getPlayers } from '../../services/players'

function Players({ user }) {
  const [players, setPlayers] = useState([])
  const [loading, setLoading] = useState(true)

  const loadPlayers = () => {
    getPlayers()
      .then(({ data }) => setPlayers(data))
      .catch((err) => console.error(err))
      .finally(() => setLoading(false))
  }

  useEffect(() => {
    loadPlayers()
  }, [])

  const handleDelete = ({ id, name }) => {
    const shouldDelete = confirm(`Are you sure you want to delete ${name}?`)

    if (shouldDelete) {
      deletePlayerById(id).then(() => loadPlayers())
    }
  }

  if (loading) return <Loading />

  return (
    <>
      <h1>Players</h1>
      <table className="resource-table">
        <thead>
          <tr>
            <th>Player Name</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          {players.map((player) => (
            <tr key={player.id}>
              <td>
                <Link to={`/players/${player.id}`}>{player.name}</Link>
              </td>
              <td>
                <Link to={`/players/${player.id}`}>
                  <button type="button" className="btn-view">
                    View
                  </button>
                </Link>
                {user && (
                  <>
                    <Link to={`/players/${player.id}/edit`}>
                      <button type="button" className="btn-edit">
                        Edit
                      </button>
                    </Link>
                    <button
                      type="button"
                      className="btn-delete"
                      onClick={() => handleDelete(player)}
                    >
                      Delete
                    </button>
                  </>
                )}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </>
  )
}

export default Players
