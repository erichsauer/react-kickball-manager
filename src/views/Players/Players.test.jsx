import { render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import Players from './Players'

it('should render a list of players', async () => {
  // MemoryRouter is needed because <Players> uses the <Link> component
  const { container } = render(
    <MemoryRouter>
      <Players />
    </MemoryRouter>
  )
  screen.getByLabelText(/loading/)
  await screen.findByText(/Ben E. Jetts/)

  expect(container).toMatchSnapshot()
})
