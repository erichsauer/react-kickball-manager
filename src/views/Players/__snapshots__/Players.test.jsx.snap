// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a list of players 1`] = `
<div>
  <h1>
    Players
  </h1>
  <table
    class="resource-table"
  >
    <thead>
      <tr>
        <th>
          Player Name
        </th>
        <th>
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <a
            href="/players/1"
          >
            Ben E. Jetts
          </a>
        </td>
        <td>
          <a
            href="/players/1"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/12"
          >
            Betty Grey
          </a>
        </td>
        <td>
          <a
            href="/players/12"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/7"
          >
            Bud E. Guy
          </a>
        </td>
        <td>
          <a
            href="/players/7"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/8"
          >
            Buster Rimes
          </a>
        </td>
        <td>
          <a
            href="/players/8"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/9"
          >
            Deb E. Downer
          </a>
        </td>
        <td>
          <a
            href="/players/9"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/13"
          >
            Dwayne Elizondo Mountain Dew Herbert Camacho
          </a>
        </td>
        <td>
          <a
            href="/players/13"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/15"
          >
            Frito Pendejo
          </a>
        </td>
        <td>
          <a
            href="/players/15"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/5"
          >
            Gary Coolman
          </a>
        </td>
        <td>
          <a
            href="/players/5"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/2"
          >
            Hans Olo
          </a>
        </td>
        <td>
          <a
            href="/players/2"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/10"
          >
            Jayne Phonda
          </a>
        </td>
        <td>
          <a
            href="/players/10"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/14"
          >
            Joe Bauers
          </a>
        </td>
        <td>
          <a
            href="/players/14"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/11"
          >
            Kara Whollot
          </a>
        </td>
        <td>
          <a
            href="/players/11"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/3"
          >
            Ron Long
          </a>
        </td>
        <td>
          <a
            href="/players/3"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/4"
          >
            Shorty Biggum
          </a>
        </td>
        <td>
          <a
            href="/players/4"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
      <tr>
        <td>
          <a
            href="/players/6"
          >
            Terry Cruise
          </a>
        </td>
        <td>
          <a
            href="/players/6"
          >
            <button
              class="btn-view"
              type="button"
            >
              View
            </button>
          </a>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
